From 4a49b71e224fc8d9ee652fdf593d094c8bc73b68 Mon Sep 17 00:00:00 2001
From: "Zhong,Fangjian" <fangjian.zhong@intel.com>
Date: Wed, 19 Jul 2023 22:12:51 +0800
Subject: [PATCH 13/54] fastboot: Erase stored rollback slots when status
 changes.

Stored rollback slots should be removed when status changes during
fastboot flashing unlock scenario.

Tracked-On: OAM-111252
Signed-off-by: Zhong,Fangjian <fangjian.zhong@intel.com>
---
 libfastboot/fastboot_flashing.c | 24 ++++++++++++++++++++++++
 1 file changed, 24 insertions(+)

diff --git a/libfastboot/fastboot_flashing.c b/libfastboot/fastboot_flashing.c
index a7af14f..917cb00 100644
--- a/libfastboot/fastboot_flashing.c
+++ b/libfastboot/fastboot_flashing.c
@@ -40,6 +40,7 @@
 #include "gpt.h"
 #include "intel_variables.h"
 #include "android.h"
+#include "tpm2_security.h"
 
 static cmdlist_t cmdlist;
 
@@ -114,6 +115,29 @@ EFI_STATUS change_device_state(enum device_state new_state, BOOLEAN interactive)
 	}
 #endif
 
+	if (UNLOCKED == new_state) {
+		info(L"Erasing rollback index...");
+
+/*
+ * Accroding to AVB_AB implementation, only 2 rollback slots will be used,
+ * we reset 2 possible slots only.
+ */
+		for (int slot = 0; slot < 2; slot++) {
+			uint64_t idx;
+#ifdef USE_TPM
+			ret = read_rollback_index_tpm2(slot, &idx);
+			if (EFI_SUCCESS == ret) {
+				ret = write_rollback_index_tpm2(slot, 0);
+			}
+#else
+			ret = read_efi_rollback_index(slot, &idx);
+			if (EFI_SUCCESS == ret) {
+				ret = write_efi_rollback_index(slot, 0);
+			}
+#endif
+		}
+	}
+
 	ret = set_current_state(new_state);
 	if (EFI_ERROR(ret)) {
 		if (interactive)
-- 
2.25.1

